linode-launch.py

```diff
@@ -5,184 +5,107 @@
-
-def create(*args, **kwargs):
-    try:
-        do_create(*args, **kwargs)
-    except Exception as e:
-        logging.exception(e)
-        os._exit(1)
-
-def launch(client):
-    datacenters = client.avail_datacenters()
-    plans = client.avail_linodeplans()
-    distributions = client.avail_distributions()
-    kernels = client.avail_kernels()
-
-    datacenter = filter(lambda d: d[u'LOCATION'].lower().find(DATACENTER.lower()) >= 0, datacenters)[0][u'DATACENTERID']
-    distribution = filter(lambda d: d[u'LABEL'].lower().find(DISTRIBUTION.lower()) >= 0, distributions)[0][u'DISTRIBUTIONID']
-    kernel = filter(lambda k: k[u'LABEL'].lower().find(str(KERNEL).lower()) >= 0, kernels)[0][u'KERNELID']
-
-    running = client.linode_list()
-
-    with closing(ThreadPool(50)) as pool:
-        for machine in CLUSTER:
-            for i in range(machine['count']):
-                logging.info("%s", "i={i}".format(i=i))
-                pool.apply_async(create, (client, running, datacenter, plans, distribution, kernel, machine, i))
-        pool.close()
-        pool.join()
-
-    logging.info("%s", client.linode_list())
+
+def launch(api4_client):
+    print('')
+    print('currently running instances:')
+    for c in api4_client.linode.instances():
+        print(c.label)
+
+    print('')
+    print('currently existing volumes:')
+    for v in api4_client.volumes():
+        print('%s size %d region %s linode %s' % (
+            v.label, v.size, v.region, v.linode_id))
+    print('')
+    sys.stdout.flush()
+
+    regions = api4_client.regions()
+    for r in regions:
+        if str(r).lower().__contains__(prm_region.lower()):
+            region = r
+            break
+    logging.info('region is %s' % region)
+
+    types = api4_client.linode.types()
+    new_linodes = {}
+    for c in CLUSTER:
+        plan = c['plan']
+        count = c['count']
+        prefix = c['prefix']
+        ansible_group = prefix + 's'
+        for t in types:
+            if str(t).lower().__contains__(plan.lower()):
+                vmtype = t
+                break
+        for host in range(0, count):
+            new_linode, password = api4_client.linode.instance_create(
+                vmtype,
+                region,
+                image=prm_image,
+                authorized_keys = ssh_public_key_file)
+            ipv6addr = new_linode.ipv6.split('/')[0]
+            label = new_linode.label
+            new_linodes[str(new_linode)] = (vmtype, ansible_group, ipv6addr, region, password)
+            logging.info('%s, %s, %s, %s' % (new_linode, ipv6addr, ansible_group, vmtype))
+            if prefix == 'osd':
+                for dev in range(0, prm_osds_per_host):
+                    new_vol = api4_client.volume_create(
+                                'v_%s_%d_v%d' % (prefix, host, dev),
+                                size=prm_osd_size,
+                                region=region)
```

Comment:
How did you figure this out?   I saw nothing in the docs or even [linode_api4 code](https://github.com/linode/linode_api4-python/blob/master/linode_api4/objects/volume.py) about it.  So the "disk" is an ephemeral virtual device on the same physical host as the VM?

The thing I hate the most about disks is that you can't control the available space except by changing your linode "plan" (what size linode you use).  Volumes seemed to solve that problem nicely, but if we double network round trips with them, that could cause problems.     You can't dynamicallly attach/remove disks on a running system, but you can do that with volumes.  I'll do some tests and see what the latency looks like for raw volume vs raw disk.

Reply:
> How did you figure this out? I saw nothing in the docs or even linode_api4 code about it. So the "disk" is an ephemeral virtual device on the same physical host as the VM?

Just reading the blogs. I don't recall a particular blog/doc which made that clear.

Reply:
You were absolutely right.  I ran fio on a volume and on a disk.  The volume got on the order of 100 IOPS and the disk got on the order of 4000 IOPS.  Could use the disk for SSD journaling since it's so much faster than the volume.  But there is plenty of space on the disk if you can pay for a plan like g6-standard-8, about 600 GB in that case, way more than you need for the "system disk".  Now I have to figure out how to do what you did with the previous API, but using linode_api4.